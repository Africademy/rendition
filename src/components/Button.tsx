import { Button, ButtonProps as GrommetButtonProps } from 'grommet';
import * as React from 'react';
import styled from 'styled-components';
import asRendition from '../asRendition';
import {
	Coloring,
	DefaultProps,
	Omit,
	RenditionSystemProps,
	ResponsiveStyle,
	Sizing,
	Theme,
} from '../common-types';
import { getColor, getColoringType, normal } from '../utils';

const getHoverEffectOverride = (
	bg: string,
	color: string,
	opacity?: string,
) => {
	return `
		&:hover:enabled,
		&:focus:enabled,
		&:active:enabled {
			box-shadow: none;
			background: ${bg};
			border-color: ${bg};
			color: ${color};
			opacity: ${opacity};
		}
	`;
};

const ButtonBase = styled(Button)`
	font-family: ${props => props.theme.titleFont};
	font-weight: ${props => normal(props)};
	height: 38px;
	white-space: nowrap;
	svg {
		color: inherit !important;
		font-size: 0.875em;
	}
	/* These rules cause consistent styling when the button is rendered as a link */
	display: inline-flex;
	align-items: center;
	justify-content: center;

	&:disabled {
		cursor: not-allowed;
	}
`;

const ColouredButton = styled(ButtonBase)`
	color: white;
	${props => getHoverEffectOverride(getColor(props, 'color', 'dark'), 'white')};
`;

const Outline = styled(ButtonBase)<{ color?: string }>`
	color: ${props => props.color || props.theme.colors.text.main};
	border-color: ${props => props.color || props.theme.colors.text.main};
	${props =>
		getHoverEffectOverride(
			props.color || props.theme.colors.text.main,
			'white',
		)};
`;

const Light = styled(ButtonBase)`
	color: ${props => props.theme.colors.text.main};
	border-color: ${props => props.theme.colors.text.main};
	background: white;
	border: none;
	${props =>
		getHoverEffectOverride('white', props.theme.colors.secondary.main, '0.9')};

	&:disabled {
		opacity: 1;
		color: #c6c8c9;
	}
`;

const LightOutline = styled(Light)`
	color: white;
	border: 1px solid white;
	background: transparent;
	${props =>
		getHoverEffectOverride('white', props.theme.colors.secondary.main, '0.9')};

	&:disabled {
		opacity: 1;
		color: #c6c8c9;
	}
`;

const Plain = styled(ButtonBase)<{ hoverColor?: string; color?: string }>`
	color: ${props => props.color || props.theme.colors.text.main};
	height: auto;
	font-weight: ${props => normal(props)};
	border-radius: 0;
	${props =>
		getHoverEffectOverride(
			'none',
			props.hoverColor || props.theme.colors.text.light,
		)};
`;

const underlineButtonActiveStyles = (props: ThemedButtonProps) => `
	color: ${props.color || props.theme.colors.text.main};
	background: none;
	box-shadow: 0px -1px 0 0px inset;
`;

const Underline = styled(Plain)<{ active?: boolean; color?: string }>`
	${props => (props.active ? underlineButtonActiveStyles(props) : '')}
	border-bottom: 1px solid;
	background: none;
	&:hover:enabled,
	&:focus:enabled,
	&:active:enabled {
		${underlineButtonActiveStyles};
	}
`;

const getStyledButton = (
	{
		outline,
		underline,
		plain,
		active,
		light,
		...originalProps
	}: ThemedButtonProps,
	isPrimary: boolean,
) => {
	if (plain) {
		return (props: GrommetButtonProps) => (
			<Plain
				{...props}
				color={getColor(originalProps, 'color', 'main')}
				hoverColor={getColor(originalProps, 'color', 'light')}
			/>
		);
	}

	if (light) {
		if (outline) {
			return LightOutline;
		}
		return Light;
	}

	if (underline) {
		return (props: GrommetButtonProps) => (
			<Underline
				{...props}
				plain={true}
				color={getColor(originalProps, 'color', 'main')}
			/>
		);
	}

	if ((outline && !active) || !isPrimary) {
		return (props: GrommetButtonProps) => (
			<Outline {...props} color={getColor(originalProps, 'color', 'main')} />
		);
	}

	return ColouredButton;
};

const Base = (props: ThemedButtonProps) => {
	const {
		outline,
		underline,
		children,
		label,
		primary,
		color,
		plain,
		active,
		light,
		...restProps
	} = props;

	// The 'primary' and 'color' props map to semantically different properties in
	// grommet, so mapping is performed here to keep the styling cleaner.
	// In grommet, 'primary' indicates that the button is filled with a solid
	// color, whereas in rendition 'primary' is a shorthand for the primary theme
	// color. In grommet 'color' refers to the fill or outline color of the
	// button, whereas in rendition it refers to the text color of the button
	let basePrimary =
		!!props.bg ||
		(!!getColoringType(props) && !outline && !underline && !props.plain);
	let baseColor = props.bg || getColor(props, 'bg', 'main');

	// Set default to 'secondary'
	const secondary = baseColor ? false : true;
	baseColor = baseColor || props.theme.colors.secondary.main;

	// If the button is active, invert the background and text colors
	if (active) {
		basePrimary = !basePrimary;
	}

	const StyledButton = getStyledButton({ secondary, ...props }, basePrimary);

	// Note that the 'plain' property is case to a Boolean to simplify Grommets
	// default behaviour when providing an icon attribute and no label
	// see: https://github.com/grommet/grommet/issues/3030
	return (
		<StyledButton
			primary={basePrimary}
			color={baseColor}
			active={active && (underline || plain)}
			{...restProps}
			plain={!!plain || !!underline}
			label={label || children}
		/>
	);
};

interface ButtonBaseProps extends Coloring, Sizing {
	width?: ResponsiveStyle;
	bg?: string;
	outline?: boolean;
	underline?: boolean;
	light?: boolean;
}

// Grommet has its own declaration for the 'onClick' attribute, so the default
// one is omitted as Grommet takes precedence here
export interface InternalButtonProps
	extends ButtonBaseProps,
		Omit<DefaultProps, 'onClick'>,
		GrommetButtonProps {
	type?: 'submit' | 'reset' | 'button';
	label?: 'string' | JSX.Element;
}

export type ButtonProps = InternalButtonProps & RenditionSystemProps;

export interface ThemedButtonProps extends ButtonProps {
	theme: Theme;
}

export default asRendition<ButtonProps>(
	Base,
	[],
	['width', 'color', 'bg'],
) as React.FunctionComponent<ButtonProps>;
