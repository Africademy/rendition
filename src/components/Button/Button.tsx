import asRendition from '../../asRendition';

import {
	Button
} from 'grommet'

export default asRendition(Button);
