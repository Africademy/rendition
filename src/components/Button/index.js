import React from 'react';
import classNames from 'classnames';

const { PropTypes } = React;

const Button = ({
  btnStyle,
  children,
  className,
  disabled,
  onClick,
  id,
} = {}) => {
  let classes = classNames('resin-control', className);

  if (btnStyle) {
    classes += ` resin-control_btn-${btnStyle}`;
  }

  return (
    <button
      onClick={onClick}
      id={id}
      className={classes}
      disabled={disabled}
    >
      {children}
    </button>
  );
};

Button.propTypes = {
  btnStyle: PropTypes.string,
  children: React.PropTypes.node,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  onClick: PropTypes.func,
};

Button.defaultProps = {
  // 'primary', 'secondary', 'dark', 'light'
  btnStyle: 'dark',
};

export default Button;
