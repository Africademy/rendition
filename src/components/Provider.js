import * as React from 'react'
import styled, { ThemeProvider } from 'styled-components'
import defaultTheme from '../theme'

const Base = styled.div`
  font-family: ${props => props.theme.font};
`

const Provider = ({ theme, ...props }) => {
  return (
    <ThemeProvider theme={theme || defaultTheme}>
      <Base {...props} />
    </ThemeProvider>
  )
}

export default Provider
