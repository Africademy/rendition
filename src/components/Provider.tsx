import * as React from 'react';
import styled, { StyledFunction, ThemeProvider } from 'styled-components';
import { DefaultProps, Theme } from '../common-types';
import defaultTheme from '../theme';

export interface ThemedBaseProps extends DefaultProps {
	theme?: Theme;
}

const Base = (styled.div as StyledFunction<DefaultProps>)`
  font-family: ${props => props.theme.font};
`;

const Provider = ({ theme, ...props }: ThemedBaseProps) => {
	return (
		<ThemeProvider theme={theme || defaultTheme}>
			<Base {...props} />
		</ThemeProvider>
	);
};

export default Provider;
