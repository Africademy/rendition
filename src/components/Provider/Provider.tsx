import {
	Grommet
} from 'grommet'
import * as React from 'react';

const Provider = ({ theme, ...props }: any) => {
	return (
		<Grommet theme={theme} {...props} />
	);
};

export default Provider;
