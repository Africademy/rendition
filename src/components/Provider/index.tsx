import { Grommet } from 'grommet';
import merge from 'lodash/merge';
import * as React from 'react';
import styled from 'styled-components';
import { DefaultProps, Theme } from '../../common-types';
import defaultTheme from '../../theme';
import { px } from '../../utils';
import { BreakpointProvider } from './BreakpointProvider';

const Base = styled(Grommet)`
	font-family: ${props => props.theme.font};
	font-size: ${props => px(props.theme.fontSizes[1])};

	h1,
	h2,
	h3,
	h4,
	h5,
	h6,
	button {
		font-family: ${props => props.theme.titleFont};
	}
`;

const Provider = ({ theme, ...props }: ThemedProvider) => {
	const providerTheme = merge(defaultTheme, theme);
	return (
		<BreakpointProvider breakpoints={providerTheme.breakpoints}>
			<Base theme={providerTheme} {...props} />
		</BreakpointProvider>
	);
};

export interface ThemedProvider extends DefaultProps {
	theme?: Partial<Theme>;
}

export default Provider;
