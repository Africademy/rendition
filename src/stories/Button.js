import React from 'react';
import { storiesOf } from '@kadira/storybook';
import { Button } from '../index';

storiesOf('Button', module)
.addWithInfo('Standard', () => {
  return (
    <div className="buttonContainer">
      <Button color="primary">primary</Button>
      <Button color="secondary">secondary</Button>
      <Button color="tertiary">tertiary</Button>
      <Button color="success">success</Button>
      <Button color="info">info</Button>
      <Button color="warning">warning</Button>
      <Button color="danger">danger</Button>
    </div>
  );
})
.addWithInfo('Outline', () => {
  return (
    <div className="buttonContainer">
      <Button outline color="primary">primary</Button>
      <Button outline color="secondary">secondary</Button>
      <Button outline color="tertiary">tertiary</Button>
      <Button outline color="success">success</Button>
      <Button outline color="info">info</Button>
      <Button outline color="warning">warning</Button>
      <Button outline color="danger">danger</Button>
    </div>
  );
})
.addWithInfo('Disabled', () => {
  return (
    <div className="buttonContainer">
      <Button disabled color="primary">primary</Button>
      <Button disabled color="secondary">secondary</Button>
      <Button disabled color="tertiary">tertiary</Button>
      <Button disabled color="success">success</Button>
      <Button disabled color="info">info</Button>
      <Button disabled color="warning">warning</Button>
      <Button disabled color="danger">danger</Button>
    </div>
  );
})
.addWithInfo('Large', () => {
  return (
    <div className="buttonContainer">
      <Button color="primary" size="lg">Large Button</Button>
      <Button color="secondary" size="lg">Large Button</Button>
      <Button color="tertiary" size="lg">Large Button</Button>
    </div>
  );
})
.addWithInfo('Small', () => {
  return (
    <div className="buttonContainer">
      <Button color="primary" size="sm">Small Button</Button>
      <Button color="secondary" size="sm">Small Button</Button>
      <Button color="tertiary" size="sm">Small Button</Button>
    </div>
  );
});
