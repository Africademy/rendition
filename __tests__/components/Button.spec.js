/* globals expect, test */
import React from 'react'
import renderer from 'react-test-renderer'
import Provider from '../../src/components/Provider'
import Button from '../../src/components/Button'

test('Button renders correctly', () => {
  const component = renderer.create(
    <Provider>
      <Button>Click me</Button>
    </Provider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Styled Button renders correctly', () => {
  const component = renderer.create(
    <Provider>
      <Button m={2}>Click me</Button>
    </Provider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Styled Anchor Button renders correctly', () => {
  const component = renderer.create(
    <Provider>
      <Button.a href={'#'} m={2}>Click me</Button.a>
    </Provider>
  )
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
