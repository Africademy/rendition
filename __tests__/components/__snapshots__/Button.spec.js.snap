// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<div
  className="sc-bdVaJa jKopLr"
>
  <button
    className="sc-VigVT hegYug sc-gqjmRU yBQjZ"
  >
    Click me
  </button>
</div>
`;

exports[`Styled Anchor Button renders correctly 1`] = `
<div
  className="sc-bdVaJa jKopLr"
>
  <a
    className="sc-jTzLTM ctytKv sc-jzJRlG fKAaUV sc-fjdhpX izaRYO"
    href="#"
  >
    Click me
  </a>
</div>
`;

exports[`Styled Button renders correctly 1`] = `
<div
  className="sc-bdVaJa jKopLr"
>
  <button
    className="sc-VigVT kKVfAK sc-gqjmRU yBQjZ"
  >
    Click me
  </button>
</div>
`;
